core/src/main/java/tc/oc/pgm/community/modules/FreezeMatchModule.java

```diff
@@ -0,0 +1,349 @@
+    if (freeze.isFrozen(event.getPlayer()) && !event.getPlayer().hasPermission(Permissions.STAFF)) {
+      boolean allow =
+          ALLOWED_CMDS.stream()
+              .filter(cmd -> event.getMessage().startsWith(cmd))
+              .findAny()
+              .isPresent();
+
+      if (!allow) {
+        // Don't allow commands except for those related to chat.
+        event.setCancelled(true);
+      }
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onPlayerMove(final PlayerMoveEvent event) {
+    if (freeze.isFrozen(event.getPlayer())) {
+      Location old = event.getFrom();
+      old.setPitch(event.getTo().getPitch());
+      old.setYaw(event.getTo().getYaw());
+      event.setTo(old);
+    }
+  }
+
+  @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
+  public void onVehicleMove(final VehicleMoveEvent event) {
+    if (!event.getVehicle().isEmpty() && freeze.isFrozen(event.getVehicle().getPassenger())) {
+      event.getVehicle().setVelocity(new Vector(0, 0, 0));
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onVehicleEnter(final VehicleEnterEvent event) {
+    if (freeze.isFrozen(event.getEntered())) {
+      event.setCancelled(true);
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onVehicleExit(final VehicleExitEvent event) {
+    if (freeze.isFrozen(event.getExited())) {
+      event.setCancelled(true);
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onBlockBreak(final BlockBreakEvent event) {
+    if (freeze.isFrozen(event.getPlayer())) {
+      event.setCancelled(true);
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onBlockPlace(final BlockPlaceEvent event) {
+    if (freeze.isFrozen(event.getPlayer())) {
+      event.setCancelled(true);
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onBucketFill(final PlayerBucketFillEvent event) {
+    if (freeze.isFrozen(event.getPlayer())) {
+      event.setCancelled(true);
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onBucketEmpty(final PlayerBucketEmptyEvent event) {
+    if (freeze.isFrozen(event.getPlayer())) {
+      event.setCancelled(true);
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW) // ignoreCancelled doesn't seem to work well here
+  public void onPlayerInteract(final PlayerInteractEvent event) {
+    if (freeze.isFrozen(event.getPlayer())) {
+      event.setCancelled(true);
+    }
+  }
+
+  @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
+  public void onInventoryClick(final InventoryClickEvent event) {
+    if (event.getWhoClicked() instanceof Player) {
+      if (freeze.isFrozen(event.getWhoClicked())) {
+        event.setCancelled(true);
+      }
+    }
+  }
```

Comment:
Same for this, why not just cancel all `InventoryEvent`s, that way to can also cancel events like `InventoryOpenEvent` and opening other interfaces like Anvils, Crafting tables..

Reply:
Good thinking, however I think it’s fine to allow the inventory open event. Let’s say a moderator is questioning the contents of the frozen player’s inventory, it may be useful to allow the user to open up their inventory and see. Plus with the inventory click blocking it, they can’t edit it anyway.

Reply:
@KingOfSquares Also just checked, InventoryEvent is not cancellable either.